practicas/Iroyoshi/P4/Práctia4_CesarCossioGuerrero.py

```diff
@@ -0,0 +1,359 @@
+        # x': concatenation of x1 and x2 embeddings   -->
+        # self.embeddings regresa un vector por cada uno de los índices que se les pase como entrada.
+        # view() les cambia el tamaño para concatenarlos
+        embeds = self.embeddings(inputs).view((-1,self.context_size * self.embedding_dim))
+        # h: tanh(W_1.x' + b)  -->
+        out = torch.tanh(self.linear1(embeds))
+        # W_2.h                 -->
+        out = self.linear2(out)
+        # log_softmax(W_2.h)      -->
+        # dim=1 para que opere sobre renglones, pues al usar batchs tenemos varios vectores de salida
+        log_probs = F.log_softmax(out, dim=1)
+
+        return log_probs
+
+
+# + id="Evob5HAROoHn"
+def preprocess_corpus(corpus: list[str]) -> list[str]:
+    """Función de preprocesamiento
+
+    Esta función está diseñada para preprocesar
+    corpus para modelos del lenguaje neuronales.
+    Agrega tokens de inicio y fin, normaliza
+    palabras a minusculas
+    """
+    preprocessed_corpus = []
+    for sent in corpus:
+        result = [word.lower() for word in sent]
+        # Al final de la oración
+        result.append("<EOS>")
+        result.insert(0, "<BOS>")
+        preprocessed_corpus.append(result)
+    return preprocessed_corpus
+
+
+# + id="2GGww6gxOlgv"
+def get_words_freqs(corpus: list[list[str]]):
+    """Calcula la frecuencia de las palabras en un corpus"""
+    words_freqs = {}
+    for sentence in corpus:
+        for word in sentence:
+            words_freqs[word] = words_freqs.get(word, 0) + 1
+    return words_freqs
+
+
+# + id="1WijHio1Oie2"
+UNK_LABEL = "<UNK>"
+def get_words_indexes(words_freqs: dict) -> dict:
+    """Calcula los indices de las palabras dadas sus frecuencias"""
+    result = {}
+    for idx, word in enumerate(words_freqs.keys()):
+        # Happax legomena happends
+        if words_freqs[word] == 1:
+            # Temp index for unknowns
+            result[UNK_LABEL] = len(words_freqs)
+        else:
+            result[word] = idx
+
+    return {word: idx for idx, word in enumerate(result.keys())}, {idx: word for idx, word in enumerate(result.keys())}
+
+
+# + id="314qLY5u6SK5"
+def get_word_id(words_indexes: dict, word: str) -> int:
+    """Obtiene el id de una palabra dada
+
+    Si no se encuentra la palabra se regresa el id
+    del token UNK
+    """
+    unk_word_id = words_indexes[UNK_LABEL]
+    return words_indexes.get(word, unk_word_id)
+
+
+# + [markdown] id="us5tCwkpyrDu"
+# ## Preparemos el **corpus** que usaremos para graficar
+
+# + id="aFeTkYgQO15H"
+corpus = preprocess_corpus(reuters.sents())
+
+# + id="dTu3C-ipO1Av"
+words_freqs = get_words_freqs(corpus)
+
+# + id="YtveA198OesR"
+words_indexes, index_to_word = get_words_indexes(words_freqs)
+
+# + [markdown] id="KmnKftLcy4dO"
+# ## Cargamos el **modelo** entrenado en clase que se encuentra en la carpeta '/content/' usada en este notebook
+
+# + id="kzBqFYidzUfu"
+PATH = "/content/model_cuda_context_2_epoch_2.dat"
```

Comment:
El modelo no se encuentra en content si no en la raíz de tu práctica. Por lo que debería quedar así:
```suggestion
PATH = "model_cuda_context_2_epoch_2.dat"
```